### DISCUSSION TITLE
Using different timeline variables in the same procedure

### DISCUSSION THREAD

**Discussion by anjiecao** [jspsych/jsPsych#955] [Q&A] [answered]
Hi, 

I'm quite confused by how the timeline variables work and I would really appreciate some suggestions!
So my goal is to create a procedure, where participants see many trials. In each trial, they see a random picture from group A followed by seeing a random picture from group B.  From what I understand, I can create a procedure like this: 

```r
 var my_procedure = {
        timeline: [

            { type: 'image-keyboard-response',
             stimulus:jsPsych.timelineVariable('picGroupA'),
    	   choices:  jsPsych.NO_KEYS},

            
          { type: 'image-keyboard-response',
        stimulus: jsPsych.timelineVariable('picGroupB'),
        choices: jsPsych.NO_KEYS}
        

        ],
        
        
        
        timeline_variables: [
            {'picGroupA': 'images/A01.jpg'},
            {'picGroupA': 'images/A02.jpg'},
            {'picGroupA': 'images/A03.jpg'},
            {'picGroupB': 'images/B01.jpg'},
            {'picGroupB': 'images/B02.jpg'},
            {'picGroupB': 'images/B03.jpg'},             
        ],
        
        randomize_order: true
        
        
        
    }
```

But I'm experiencing a lot of bizarre behaviors with this approach, such as can't find the stimulus etc. I think I misunderstand how timeline_variables work....any input will be greatly appreciated!

**Comment by anjiecao** [comment, marked as answer]
I think I figured something out  -- not sure if it is the best practice, but I'll just share it here and in case. 
I can escape the problem of timeline variables by using a function: 

So instead of 

```js
{ 
	type: 'image-keyboard-response',
        stimulus: jsPsych.timelineVariable('picGroupB'),
        choices: jsPsych.NO_KEYS
}
```

I can have 

```js
{ 
	type: 'image-keyboard-response',
        stimulus: function(){
                return jsPsych.randomization.sampleWithoutReplacement(['images/B01.jpg', 'images/B02.jpg', 'images/B03.jpg'], 1)[0];
        },
        choices: jsPsych.NO_KEYS
}
```

Doing so can also help achieve "select random one from the picture group. I'm not sure if it is the best practice so I'll leave this question open for now.

**Reply by jodeleeuw** [reply]
If the goal is just to select one from a random group then I think your approach is the strategy I would recommend. 

If you wanted to do it with timeline variables it would look a bit different. Timeline variables are defined in sets. Each object must contain all of the variables for one run through the timeline. In your case you'd need to do something like this:

```js
timeline_variables: [
        {'picGroupA': 'images/A01.jpg', 'picGroupB': 'images/B01.jpg'},
        {'picGroupA': 'images/A02.jpg', 'picGroupB': 'images/B01.jpg'},
        {'picGroupA': 'images/A03.jpg', 'picGroupB': 'images/B01.jpg'},
        {'picGroupA': 'images/A01.jpg', 'picGroupB': 'images/B02.jpg'},
        {'picGroupA': 'images/A02.jpg', 'picGroupB': 'images/B02.jpg'},
        {'picGroupA': 'images/A03.jpg', 'picGroupB': 'images/B02.jpg'},
	{'picGroupA': 'images/A01.jpg', 'picGroupB': 'images/B03.jpg'},
        {'picGroupA': 'images/A02.jpg', 'picGroupB': 'images/B03.jpg'},
        {'picGroupA': 'images/A03.jpg', 'picGroupB': 'images/B03.jpg'},    
],
```

This would create all 9 possibilities, and then you could sample randomly from them by either setting `randomize_order: true`  or using a `sample` method.
